caikit/runtime/http_server.py

```diff
@@ -0,0 +1,460 @@
+from typing import Iterable, Optional, Type, Union, get_args, get_origin
+import asyncio
+import json
+import re
+import ssl
+import threading
+import time
+
+# Third Party
+from fastapi import FastAPI, Request, Response
+from grpc import StatusCode
+from sse_starlette import EventSourceResponse, ServerSentEvent
+import numpy as np
+import pydantic
+import uvicorn
+
+# First Party
+from py_to_proto.dataclass_to_proto import Annotated  # Imported here for 3.8 compat
+import aconfig
+import alog
+
+# Local
+from caikit.config import get_config
+from caikit.core.data_model import DataBase
+from caikit.core.data_model.dataobject import make_dataobject
+from caikit.core.toolkit.sync_to_async import async_wrap_iter
+from caikit.runtime.server_base import RuntimeServerBase
+from caikit.runtime.service_factory import ServicePackage, ServicePackageFactory
+from caikit.runtime.service_generation.rpcs import CaikitRPCBase
+from caikit.runtime.servicers.global_predict_servicer import GlobalPredictServicer
+from caikit.runtime.types.caikit_runtime_exception import CaikitRuntimeException
+import caikit.core.toolkit.logging
+
+## Globals #####################################################################
+
+log = alog.use_channel("HTTP")
+
+# Registry of DM -> Pydantic model mapping to avoid errors when reusing messages
+# across endpoings
+PYDANTIC_REGISTRY = {}
+
+
+# Mapping from GRPC codes to their corresponding HTTP codes
+# pylint: disable=line-too-long
+# CITE: https://chromium.googlesource.com/external/github.com/grpc/grpc/+/refs/tags/v1.21.4-pre1/doc/statuscodes.md
+GRPC_CODE_TO_HTTP = {
+    StatusCode.OK: 200,
+    StatusCode.INVALID_ARGUMENT: 400,
+    StatusCode.FAILED_PRECONDITION: 400,
+    StatusCode.OUT_OF_RANGE: 400,
+    StatusCode.UNAUTHENTICATED: 401,
+    StatusCode.PERMISSION_DENIED: 403,
+    StatusCode.NOT_FOUND: 404,
+    StatusCode.ALREADY_EXISTS: 409,
+    StatusCode.ABORTED: 409,
+    StatusCode.RESOURCE_EXHAUSTED: 429,
+    StatusCode.CANCELLED: 499,
+    StatusCode.UNKNOWN: 500,
+    StatusCode.DATA_LOSS: 500,
+    StatusCode.UNIMPLEMENTED: 501,
+    StatusCode.UNAVAILABLE: 501,
+    StatusCode.DEADLINE_EXCEEDED: 504,
+}
+
+
+# These keys are used to define the logical sections of the request and response
+# data structures.
+REQUIRED_INPUTS_KEY = "inputs"
+OPTIONAL_INPUTS_KEY = "parameters"
+
+## RuntimeHTTPServer ###########################################################
+
+
+class RuntimeHTTPServer(RuntimeServerBase):
+    """An implementation of a FastAPI server that serves caikit runtimes"""
+
+    ###############
+    ## Interface ##
+    ###############
+
+    def __init__(self, tls_config_override: Optional[aconfig.Config] = None):
+        super().__init__(get_config().runtime.http.port, tls_config_override)
+        # if not set in config, timeout is None and unvicorn accepts None or number of seconds
+        unvicorn_timeout_graceful_shutdown = get_config().runtime.http.timeout
```

Comment:
NIT: This should live down next to the place where it's used. Having it up here makes it difficult to trace how it's used